File: src/Belajar_SOLID_02_OOP_Inheritance/Java/Hierarchical_Inheritance.java
package Belajar_SOLID_02_OOP_Inheritance.Java;

// Sama seperti penerapan inheritance pada umumnya,
// jenis ini merupakan sebuah SuperClass yang diwarisi oleh beberapa SubClass

class Hewan{
    public void walk(){
        System.out.println(getClass().getSimpleName() + " walk!");
    }
}

class Carnivora extends Hewan{
    public void eat(){
        System.out.println(getClass().getSimpleName() + " eat!");
    }
}

class Kucing extends Carnivora{}

class Singa extends Carnivora{}

public class Hierarchical_Inheritance {
    public static void main(String[] args) {
        Kucing neko = new Kucing();
        neko.eat();
        neko.walk();
    }
}

File: src/Belajar_SOLID_02_OOP_Inheritance/Kotlin/Hierarchical_Inheritance.kt
package Belajar_SOLID_02_OOP_Inheritance.Kotlin

//Sama seperti penerapan inheritance pada umumnya, jenis ini merupakan sebuah SuperClass yang diwarisi oleh beberapa SubClass
open class Hewan{
    fun walk(){
        println("${javaClass.simpleName} walk! ")
    }
}

open class Carnivora:Hewan(){
    fun eat(){
        println("${javaClass.simpleName} eat!")
    }
}
open class Herbivora:Hewan(){
    fun eat(){
        println("${javaClass.simpleName} eat!")
    }
}

class Kucing:Carnivora()

class Singa:Carnivora()

class Sapi:Herbivora()

fun main(){
    val kucing = Kucing()
    kucing.eat()
    kucing.walk()

    val singa = Singa()
    singa.eat()
    singa.walk()
}